src/main/java/org/truffleruby/language/arguments/keywords/NonEmptyKeywordDescriptor.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.language.arguments.keywords;
+
+import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import org.truffleruby.RubyLanguage;
+import org.truffleruby.core.symbol.RubySymbol;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+public class NonEmptyKeywordDescriptor implements KeywordDescriptor {
+
+    @CompilationFinal(dimensions = 1) private final String[] keywords;
+    private final boolean alsoSplat;
+    private final int hashIndex;
+
+    @CompilationFinal(dimensions = 1) private final RubySymbol[] keywordSymbols;
+
+    public NonEmptyKeywordDescriptor(RubyLanguage language, String[] keywords, boolean alsoSplat, int hashIndex) {
+        this.keywords = keywords;
+        this.alsoSplat = alsoSplat;
+        this.hashIndex = hashIndex;
+        keywordSymbols = Arrays.stream(keywords).map(language::getSymbol).toArray(n -> new RubySymbol[n]);
+    }
+
+    public String[] getKeywords() {
+        return keywords;
+    }
+
+    public boolean isAlsoSplat() {
```

Comment:
So this returns whether the kwargs are in pairs/splat directly in the frame arguments, or as a RubyHash, right?
Would be good to document.

What does the 'also' mean here? I thought it's about rest+kwargs or so.

Reply:
'As well as the flattened keyword argument values described by the descriptor, there is _also_ a hash that needs to be taken into account, and it's there because someone splatted it in when calling.'

Reply:
Oh, so for `h={a:1,b:2}; foo(**h)`, in the frame args there is `[...., h, :a, 1, :b 2]` and so the caller also "splats" the pairs in the frame args in that case?
That's a neat idea, definitely needs documentation on this class though :)